maven-caching/src/main/java/org/apache/maven/caching/CacheControllerImpl.java

```diff
@@ -397,67 +415,126 @@ private RestoredArtifact restoredArtifact( org.apache.maven.artifact.Artifact pa
+        String fullManifest = IOUtils.toString( jarFile.getInputStream( entry ), StandardCharsets.UTF_8.name() );
+        String modified = fullManifest.replaceAll( toReplace, replacement );
 
-        final ArtifactHandler artifactHandler = artifactHandlerManager.getArtifactHandler( artifact.getType() );
-        String packaging = artifactHandler.getPackaging();
-        if ( !Objects.equals( packaging, artifact.getType() ) )
-        {
-            artifactDigestByKey.put( KeyUtils.getArtifactKey( artifact, packaging ), projectArtifact );
-        }
-        String extension = artifactHandler.getExtension();
-        if ( !Objects.equals( extension, artifact.getType() ) && !Objects.equals( extension, packaging ) )
+        byte[] bytes = modified.getBytes( StandardCharsets.UTF_8 );
+        JarEntry newEntry = new JarEntry( entry.getName() );
+        jos.putNextEntry( newEntry );
+        jos.write( bytes );
+    }
+
+    private static boolean isArchive( File file )
+    {
+        String fileName = file.getName();
+        if ( !file.isFile() || file.isHidden() )
         {
-            artifactDigestByKey.put( KeyUtils.getArtifactKey( artifact, extension ), projectArtifact );
+            return false;
         }
+        return fileName.endsWith( ".jar" ) || fileName.endsWith( ".zip" )
+                || fileName.endsWith( ".war" ) || fileName.endsWith( ".ear" );
     }
 
-    private ProjectsInputInfo calculateInput( MavenProject project, MavenSession session )
+    private Future<File> createDownloadTask( CacheResult cacheResult, CacheContext context, MavenProject project,
+                                             Artifact artifact, String originalVersion )
     {
-        try
+        final FutureTask<File> downloadTask = new FutureTask<>( () ->
         {
-            if ( projectIndex == null )
+            LOGGER.debug( "Downloading artifact {}", artifact.getArtifactId() );
+            final Path artifactFile = localCache.getArtifactFile( context, cacheResult.getSource(),
+                    artifact );
+            if ( !Files.exists( artifactFile ) )
             {
-                projectIndex = session.getProjects().stream()
-                        .collect( Collectors.toMap( BuilderCommon::getKey, Function.identity() ) );
+                throw new FileNotFoundException(
+                        "Missing file for cached build, cannot restore. File: " + artifactFile );
             }
-            final MavenProjectInput inputs = new MavenProjectInput( project, projectIndex, session, cacheConfig,
-                    artifactDigestByKey, repoSystem, localCache, remoteCache );
-            return inputs.calculateChecksum();
-        }
-        catch ( Exception e )
+            LOGGER.debug( "Downloaded artifact " + artifact.getArtifactId() + " to: " + artifactFile );
+            return adjustArchiveArtifactVersion( project, originalVersion, artifactFile ).toFile();
```

Comment:
will be good to keep under flag - performance penalty of copying could be significant

Reply:
> will be good to keep it under flag - performance penalty of copying/repacking could be significant and not necessary in verify scenarios

done